File: decompiler/dec.clj
(require  '[clojure.java.io :as io])

;;把整数转成十六进制字符串
(defn toHexString [num]
	(.toUpperCase (java.lang.Integer/toHexString num)))

;;取魔数校验文件格式
(defn getMagicNumber [in]
	(loop [magic ""
		      counter 4]
		      (if (> counter 0)  
		      	(recur (str magic (toHexString (.readUnsignedByte in))) 
		      		(dec counter))
		      	magic)
		)
)

;;检查文件是否是class文件，前4个字节为魔数0xCAFEBABE，并打印版本号
(defn check-class-file [in]
	(when (= (getMagicNumber in) "CAFEBABE")
			 (println "This is a JVM class file") 
			 true)
)

;;取版本号
(defn getVersion [in]
	(let [minor (.readUnsignedShort in)
		  major (.readUnsignedShort in)]
		  (str major "." minor)))

;;对class文件进行反编译
(defn decompiler [filename]
	(with-open [in (java.io.DataInputStream. (java.io.FileInputStream. filename))]
		(when (check-class-file in)
			 (println (str "Version is " (getVersion in)))
			 )


	)
)


;;执行
(decompiler "d:/work/jvm/testfiles/Helloworld.class")

File: decompiler/Dec.java
import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.IOException;

/**
 * 对java class文件进行反编译
 * @author xinbo.zhangxb
 *
 */
public class Dec {
	private static final String magicNum = "CAFEBABE";

	/**
	 * @param args
	 * @throws IOException
	 */
	public static void main(String[] args) throws IOException {
		String filename = "d:\\work\\jvm\\testfiles\\Helloworld.class";
		DataInputStream in = new DataInputStream(new FileInputStream(filename));
		
		//magic number 
		int b = 0;
		String magic = "";
		for (int i = 0; i < 4; i++) {
			b = in.readUnsignedByte();
			magic += Integer.toHexString(b).toUpperCase();
		}

		if (magicNum.equals(magic)) {
			System.out.println("This is a JVM class file");
		}

		//version
		int minorVersion = in.readUnsignedShort();
		int majorVersion = in.readUnsignedShort();
		System.out.println("class version:" + majorVersion + "." + minorVersion);
		in.close();

	}

}
